pandas/core/dtypes/concat.py

```diff
@@ -193,7 +193,8 @@ def _concat_categorical(to_concat, axis=0):
 
     def _concat_asobject(to_concat):
         to_concat = [x.get_values() if is_categorical_dtype(x.dtype)
-                     else np.asarray(x).ravel() for x in to_concat]
+                     else np.asarray(x).ravel() if not is_datetime64tz_dtype(x)
```

Comment:
this is not the way
_concat_conpat already handles this

Reply:
@jreback after closer examination, I'm not sure if  _concat_compat actually already handles this. If the type is also a category, it simply calls _concat_categorical which then summons _concat_asobject. I think that therefore the change to the code should happen inside _concat_categorical or _concat_asobject.

Reply:
I guess I'm not sure why we aren't doing

```python
to_concat = [np.asarray(x.astype(object)) for x in to_concat]
```

The function name is `_concat_asobject`, so let's convert to object and concat them. Can you see if that works?

Please write the test first though, and ensure that the test fails with master. Then implement the fix and make sure it works.

Reply:
Seems like we have some inconsistencies in `concat` that's breaking this approach. IIRC there's an outstanding PR fixing some of these up.

I was going by the loose rule of "different dtypes means the result type is object". But things like 

```python
In [1]: import pandas as pd; import numpy as np; import pickle

In [2]: a = pd.Series([1, 2], dtype='category')

In [3]: b = pd.Series([1, None], dtype='category')

In [4]: pd.concat([a, b], ignore_index=True)
Out[4]:
0    1.0
1    2.0
2    1.0
3    NaN
dtype: float64
```

mess that up. Not sure what's best here...